server/testkit/tests/mtls_test.rs

```diff
@@ -0,0 +1,277 @@
+enum TestCase {
+    Valid,
+    ServerCertSanInvalid,
+    ServerWithoutClientCa,
+    ClientWithoutClientCert,
+    ClientWithoutServerCa,
+}
+
+#[tokio::test]
+async fn test_mtls_basic_auth() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::Valid).await;
+
+    SslStream::connect(Pin::new(&mut tlsstream)).await.unwrap();
+
+    trace!("Waiting on listener ...");
+    let result = handle.await.expect("Failed to stop task.");
+
+    // If this isn't true, it means something failed in the server accept process.
+    assert!(result.is_ok());
+}
+
+#[tokio::test]
+async fn test_mtls_server_san_invalid() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::ServerCertSanInvalid).await;
+
+    let err: openssl::ssl::Error = SslStream::connect(Pin::new(&mut tlsstream))
+        .await
+        .unwrap_err();
+    trace!(?err);
+    // Certification Verification Failure
+    let ossl_err = err.ssl_error().and_then(|e| e.errors().get(0)).unwrap();
+    assert!(ossl_err.code() == 167772294);
+
+    trace!("Waiting on listener ...");
+    let result = handle.await.expect("Failed to stop task.");
+
+    // Must be FALSE server should not have accepted the connection.
+    trace!(?result);
+    // SSL Read bytes (client disconnected)
+    assert!(matches!(result, Err(167773202)));
+}
+
+#[tokio::test]
+async fn test_mtls_server_without_client_ca() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::ServerWithoutClientCa).await;
+
+    // The client isn't the one that errors, the server does.
+    SslStream::connect(Pin::new(&mut tlsstream)).await.unwrap();
+
+    trace!("Waiting on listener ...");
+    let result = handle.await.expect("Failed to stop task.");
+
+    // Must be FALSE server should not have accepted the connection.
+    trace!(?result);
+    // Certification Verification Failure
+    assert!(matches!(result, Err(167772294)));
+}
+
+#[tokio::test]
+async fn test_mtls_client_without_client_cert() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::ClientWithoutClientCert).await;
+
+    // The client isn't the one that errors, the server does.
+    SslStream::connect(Pin::new(&mut tlsstream)).await.unwrap();
+
+    trace!("Waiting on listener ...");
+    let result = handle.await.expect("Failed to stop task.");
+
+    // Must be FALSE server should not have accepted the connection.
+    trace!(?result);
+    // Peer Did Not Provide Certificate
+    assert!(matches!(result, Err(167772359)));
+}
+
+#[tokio::test]
+async fn test_mtls_client_without_server_ca() {
+    let (mut tlsstream, handle, _tx) = setup_mtls_test(TestCase::ClientWithoutServerCa).await;
+
+    let err: openssl::ssl::Error = SslStream::connect(Pin::new(&mut tlsstream))
+        .await
+        .unwrap_err();
+    trace!(?err);
+    // Tls Post Process Certificate (Certificate Verify Failed)
+    let ossl_err = err.ssl_error().and_then(|e| e.errors().get(0)).unwrap();
+    assert!(ossl_err.code() == 2147483650);
```

Comment:
This test is unreliable as the error codes depend on library versions etc (both my machine and CI threw different error codes)
```suggestion
    assert!(ossl_err.reason() == Some("certificate verify failed"));
```

Reply:
:O 
I am shoock. I tried to do the code to be *extremely* specific about the error, but it's wild that the code changed between different versions of the lib.

Reply:
Yeah I'm digging into how/why it changes, but it's definitely a thing.